substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/configure/ConfigurationParser.java

```diff
@@ -61,24 +65,36 @@ public void parseAndRegister(Path path) throws IOException {
         throw new JSONParserException(errorMessage);
     }
 
-    protected static void checkAttributes(Map<String, Object> map, String type, Collection<String> requiredAttrs, Collection<String> optionalAttrs) {
-        List<String> unseenRequired = new ArrayList<>(requiredAttrs);
-        for (String key : map.keySet()) {
-            boolean required = unseenRequired.remove(key);
-            if (!required && !optionalAttrs.contains(key)) {
-                List<String> supported = new ArrayList<>(requiredAttrs);
-                supported.addAll(optionalAttrs);
-                throw new JSONParserException("Unknown attribute '" + key + "' (supported attributes: " + supported + ") in resource definition");
+    protected void checkAttributes(Map<String, Object> map, String type, Collection<String> requiredAttrs, Collection<String> optionalAttrs) {
+        Set<String> unseenRequired = new HashSet<>(requiredAttrs);
+        unseenRequired.removeAll(map.keySet());
+        if (!unseenRequired.isEmpty()) {
+            throw new JSONParserException("Missing attribute(s) [" + String.join(", ", unseenRequired) + "] in " + type);
+        }
+        Set<String> unknownAttributes = new HashSet<>(map.keySet());
+        unknownAttributes.removeAll(requiredAttrs);
+        unknownAttributes.removeAll(optionalAttrs);
 
-            }
+        if (seenUnknownAttributes.containsKey(type)) {
+            unknownAttributes.removeAll(seenUnknownAttributes.get(type));
         }
-        if (!unseenRequired.isEmpty()) {
-            throw new JSONParserException("Missing attribute '" + unseenRequired.get(0) + "' in " + type);
+
+        if (unknownAttributes.size() > 0) {
+            String message = "WARNING: Unknown attribute(s) [" + String.join(", ", unknownAttributes) + "] in " + type;
+            if (ConfigurationFiles.Options.StrictConfig.getValue()) {
```

Comment:
We should enable the strict mode via a parser's constructor parameter because for loading configurations in the agent or `native-image-configure` tool for merging, we definitely don't want to work on a configuration we don't understand and always need to fail.

Reply:
I agree, I've reworked it as described - we will now always fail fatally on unknown attributes in the agent/configuration tool